utils/release.ts

```diff
@@ -0,0 +1,161 @@
+    // Local env should use staging API token
+    NR_API_TOKEN = process.env['NR_API_TOKEN_STAGING'];
+    // Allow self-signed certs against localhost
+    process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
+  }
+
+  if (!NR_API_TOKEN) {
+    stepMessage(`No API token found for ${ENVIRONMENT}.`);
+    NR_API_TOKEN = await password({
+      message: `What is your NR API token for ${ENVIRONMENT}?`,
+    });
+
+    appendFileSync(
+      '.env',
+      `\nNR_API_TOKEN_${ENVIRONMENT.toUpperCase()}=${NR_API_TOKEN}`
+    );
+    console.log(
+      `Your NR API token for ${ENVIRONMENT} has been saved to the .env file. Do not commit this file.`
+    );
+  }
+
+  if (!GH_TOKEN) {
+    stepMessage(
+      "No GitHub token found. You may experience rate-limiting from the GitHub API if you don't enter one."
+    );
+    GH_TOKEN = await password({
+      message: 'What is your GitHub token? (optional)',
+    });
+
+    if (GH_TOKEN?.length) {
+      appendFileSync('.env', `\nGH_TOKEN=${GH_TOKEN}`);
+      console.log(
+        'Your GitHub token has been saved to the .env file. Do not commit this file.'
+      );
+    }
+  }
+
+  divider();
+  const PR_NUMBER = await input({
+    message:
+      'What is the PR number for the merge of the release branch into main?',
+  });
+  divider();
+
+  const PR_URL = generatePrUrl(PR_NUMBER);
+
+  // Set required env vars that can't be passed into function calls
+  process.env.NR_API_URL = API_ENDPOINTS[ENVIRONMENT];
+  process.env.NR_API_TOKEN = NR_API_TOKEN;
+
+  return {
+    ENVIRONMENT,
+    GH_TOKEN,
+    NR_API_TOKEN,
+    PR_NUMBER,
+    PR_URL,
+    API_ENDPOINT: API_ENDPOINTS[ENVIRONMENT],
+  };
+};
+
+const main = async () => {
+  const { ENVIRONMENT, GH_TOKEN, NR_API_TOKEN, PR_URL } = await bootstrap();
+
+  stepMessage(`Performing dry run release to ${ENVIRONMENT}...`);
+
+  // Dry run
+  const [dryRunSucceeded, dryRunFailures] = await runTasks(
+    PR_URL,
+    GH_TOKEN,
+    true
+  );
+
+  if (dryRunSucceeded) {
+    const shouldContinue = await confirm({
+      message: 'Dry run was successful. Proceed to deploy?',
+    });
+
+    if (shouldContinue) {
+      stepMessage(`🚀 Releasing to ${ENVIRONMENT}!`);
+      // Real deploy
+      runTasks(PR_URL, GH_TOKEN, false);
+    } else {
+      console.log('Aborting.');
+      process.exit(0);
+    }
+  }
+};
+
+const runTasks = async (url: string, token: string, dryRun = true) => {
+  const failures: Record<string, boolean> = {};
+
+  const allSuccess = () =>
+    Object.values(failures).every((status) => status === false);
+
+  failures['dataSources'] = await createValidateDataSources(url, token, dryRun);
+  failures['quickstarts'] = await createValidateQuickstarts(url, token, dryRun);
+
+  if (!dryRun) {
+    failures['dashboardsRequiredDatasources'] =
+      await setDashboardsRequiredDataSources(url, token);
+
+    failures['alertPoliciesRequiredDatasources'] =
+      await setAlertPoliciesRequiredDataSources(url, token);
```

Comment:
In the pipelines, a single failure in any of these would stop the subsequent ones from running. Do you think that's worth maintaining here, or are we covered with the initial dry run?

Reply:
@aswanson-nr Yeah, true. I'm gonna remove this check here since it'll crash and burn anyway.